src/execution/window_executor.cpp

```diff
@@ -1190,23 +1398,32 @@ WindowFirstValueExecutor::WindowFirstValueExecutor(BoundWindowExpression &wexpr,
 
 void WindowFirstValueExecutor::EvaluateInternal(WindowExecutorState &lstate, Vector &result, idx_t count,
                                                 idx_t row_idx) const {
-	auto &lbstate = lstate.Cast<WindowExecutorBoundsState>();
-	auto window_begin = FlatVector::GetData<const idx_t>(lbstate.bounds.data[WINDOW_BEGIN]);
-	auto window_end = FlatVector::GetData<const idx_t>(lbstate.bounds.data[WINDOW_END]);
+	auto &lvstate = lstate.Cast<WindowValueState>();
+	auto window_begin = FlatVector::GetData<const idx_t>(lvstate.bounds.data[WINDOW_BEGIN]);
+	auto window_end = FlatVector::GetData<const idx_t>(lvstate.bounds.data[WINDOW_END]);
 	auto &rmask = FlatVector::Validity(result);
 	for (idx_t i = 0; i < count; ++i, ++row_idx) {
+
+		if (lvstate.exclusion_filter) {
```

Comment:
I'm a little concerned there are performance implications for recomputing the mask on each row. We can leave it for now as being correct, but it might be nice to revisit it as some point.

Reply:
Right. At least it currently recomputes it only when necessary (at peer boundaries) with `EXCLUDE GROUP`....